explainability/explainability-core/src/main/java/org/kie/kogito/explainability/utils/WeightedLinearRegression.java

```
@@ -0,0 +1,288 @@
+            try {
+                 xInv = invertSquareMatrix(X);
+                inversionSuccessful = true;
+                this.X = xInv;
+            } catch(ArithmeticException e){
+                // if the inversion is unsuccessful, we can try slightly jittering the matrix.
+                // this will reduce the accuracy of the regression marginally, but ensures that we get results
+                jitterMatrix(X, .0000001);
+                jitterTries += 1;
+            }
+        }
+
+        // if jittering didn't work, throw an error
+        if (! inversionSuccessful){
+            throw new ArithmeticException(
+                    "Weighted Linear Regression: Matrix cannot be inverted! " +
+                            "This can be caused by a very under-specified model, where " +
+                            "the ratio of samples to features is roughly less than 0.10. This model has a ratio of " +
+                            (double) this.nsamples/this.nfeatures +
+                            ".");
+        }
+
+        // recover the coefficients by multiplying the inverse coefficient matrix by B
+        for (int i = 0; i < this.nfeatures; i++) {
+            this.C[i] = 0;
+            for (int j = 0; j < this.nfeatures; j++) {
+                this.C[i] += this.X[i][j] * this.B[j];
+            }
+        }
+
+        // mark the model as being fit and return coefficients
+        this.fitPerformed = true;
+        return this.C;
+    }
+
+
+    // MODEL METRICS ===================================================================================================
+    /**
+     * Recover the goodness-of-fit of the WLR model. This is the coefficient of determination, as per:
+     * https://en.wikipedia.org/wiki/Multiple_correlation
+     * @return the coefficient of determination
+     */
+    public Double getGoodnessOfFit(){
+        if (! this.fitPerformed) {
+            return Double.NaN;
```

Personally, I wouldn't return a `Double.NaN`. I would just throw an exception, it wouldn't escape unnoticed.

---

Yeah, there was some reason I was doing that but I've changed how the method works so I definitely need to get rid of return NaN.

---

In general I think we can/should avoid this problem entirely splitting this class in two different objects:
- `WeightedLinearRegressionResult` with all `final` fields (because the result is immutable) with the result (+ intermediate results) of `fit` procedure
- `WeightedLinearRegression` with private constructor and only static methods: `fit` that accept also `intercept` param (or two methods `fit` and `fitWithIntercept`) + the other existing methods with `WeightedLinearRegressionResult` as param

Wdyt?